### DISCUSSION TITLE
Proposal: Reorganize codebase to cleanly separate frontend and backend, TS and Go

### DISCUSSION THREAD

**Discussion by joeblubaugh** [grafana/grafana#47369] [Platform]
### TL; DR: Cleanly separate Go and TypeScript code within top-level directories

The top-level grafana directory list currently looks like (some files and directories omitted for brevity):
```zsh

├── Makefile
├── api-extractor.json
├── babel.config.json
├── build.go
├── conf
│   ├── defaults.ini
│   ├── ldap.toml
│   ├── ldap_multiple.toml
│   ├── provisioning
│   └── sample.ini
├── contribute
├── cue
│   └── scuemata
├── cue.mod
│   └── module.cue
├── data
├── devenv
├── e2e
│   ├── benchmarks
│   ├── dashboards
│   ├── dashboards-suite
│   ├── panels-suite
│   ├── run-suite
│   ├── shared
│   ├── smoke-tests-suite
│   ├── start-and-run-suite
│   ├── various-suite
│   ├── verify
│   └── verify-release
├── emails
│   ├── README.md
│   ├── assets
│   ├── grunt
│   ├── gruntfile.js
│   ├── package.json
│   └── templates
├── embed.go
├── go.mod
├── go.sum
├── grafana-mixin
├── jest.config.js
├── latest.json
├── lerna.json
├── metadata.md
├── node_modules
├── package.json
├── packages
│   ├── README.md
│   ├── grafana-data
│   ├── grafana-e2e
│   ├── grafana-e2e-selectors
│   ├── grafana-runtime
│   ├── grafana-schema
│   ├── grafana-toolkit
│   ├── grafana-ui
│   └── jaeger-ui-components
├── packaging
│   ├── conf
│   ├── deb
│   ├── docker
│   ├── mac
│   ├── publish
│   ├── rpm
│   └── wrappers
├── pkg
│   ├── README.md
│   ├── api
│   ├── build
│   ├── bus
│   ├── cmd
│   ├── components
│   ├── events
│   ├── expr
│   ├── extensions
│   ├── ifaces
│   ├── infra
│   ├── login
│   ├── middleware
│   ├── mocks
│   ├── models
│   ├── plugins
│   ├── registry
│   ├── ruleguard.rules.go
│   ├── schema
│   ├── server
│   ├── services
│   ├── setting
│   ├── tests
│   ├── tsdb
│   ├── util
│   └── web
├── plugins-bundled
│   ├── README.md
│   ├── external.json
│   └── internal
├── public
│   ├── api-merged.json
│   ├── api-spec.json
│   ├── app
│   ├── build
│   ├── dashboards
│   ├── emails
│   ├── fonts
│   ├── gazetteer
│   ├── img
│   ├── lib
│   ├── locales
│   ├── maps
│   ├── robots.txt
│   ├── sass
│   ├── test
│   ├── testdata
│   ├── vendor
│   └── views
├── scripts
├── stylelint.config.js
├── tools
├── tsconfig.json
├── tsconfig.tsbuildinfo
└── yarn.lock
```

This is a consequence of the organic growth of Grafana. In the early days there was no backend to speak of, and we've added more and more functionality over the years. As a result, there's a lot of disorganization in the root directory that makes it annoying to work with and confusing to learn.

Here's a proposed change to the directory structure:
1. Add top-level directories of `backend` and `frontend`. (Alternatively, `go` and `typescript`)
2. Move all Go code within `backend`. Don't change the directory structure otherwise.
3. Move all typescript code and tooling inside `frontend`
```
backend
|- go.mod
|- go.sum
|- build.go
|- pkg
| |- tsdb, etc
frontend
| |- tsconfig.json
| |- yarn.lock
| |- package.json
| |- public
| |   |- api-spec.json, ect
| |- packages
```

The key idea here is that the only build and run related file at the root directory is the Makefile. All the README.md etc stay where they are.

## Benefits
1. It's easier to get familiar with the two large codebases in the project. In particular pkg, packages, and public have too much conceptual overlap and should be more separated. The build tooling for the two languages and programs can also be separated. We expect an increasing number of people to contribute to the codebase. (Grafana is hiring! [Come work with us!](https://grafana.com/about/careers/)), and a well-organized one helps smooth ramp-up and day-to-day workflow.
2. It provides a clean example for other tools or programs we might add to the codebase in future.

## Costs
It's more work than it looks like. A variety of tools within the codebase and *outside* the codebase will likely need some changes to support the change in structure.

Our automatic backport system may struggle with handling moving all these files around. For some period of time, Grafanistas will have to do more manual backporting than we want to.

The work can be delivered in two large chunks, one for each top-level directory.

I'd propose doing this as part of a major rev like Grafana 9, where larger changes in the programs are sort of expected.

**Comment by jtheory** [comment]
The end goal here sounds good to me! It is confusing that we have 2 pretty independent codebases that are living together.

I'd look at the DX of getting started, and a normal dev loop that involves updating FE and/or BE, running tests, and manual local testing, and take care we don't add complexity/awkwardness to that. It's really nice to just clone the repo then run a few commands from the directory to build the FE, BE, and run a local instance

Trying to look ahead to how painful this would be....

Definitely part is updating many steps in CI, linter configs, that kind of thing. Though they'll be obviously broken, and once a thing works, we're done there.

Developer/contributor experience:
* Every single open PR will be broken, and resolving the merge conflicts will be hard, right? I've not hit one of these in recently (e.g. many commits in a branch, all of them touching files that were moved in the target branch)... but I recall they get painful.
* All links to source (or e.g. links to docs hosted in GH) will break; I don't know how widespread these are; some are probably outside our control

Relatively high pain level -- would it help to do this in more small steps? Or does that make it worse? :sweat_smile:

**Comment by joeblubaugh** [comment]
I spent a few hours fiddling with this. Even just locally, there are a wide variety of tools that all have their own local constant that tells them where the code is. I think a round of careful documentation would pay higher dividends than the amount of careful tweaking I think this would require.

**Reply by jtheory** [reply]
That makes sense (though it's also a little sad - I really appreciate clear code structure).
There might be specific top-level directories we can rename so it's clearer which side they're on -- e.g. we have `packages` + `packaging` + `pkg`... can any of these be renamed without much pain?

Is `scripts` a coherent top level dir?

...you're right, though, it would also be valuable to even have good readmes in each top level dir.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
